classes/local/datasource/images.php

```diff
@@ -0,0 +1,81 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace mod_cms\local\datasource;
+
+use stdClass;
+use context_system;
+
+/**
+ * Data source for images defined in the template.
+ *
+ * @package   mod_cms
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2023, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class site extends base {
+    /** The maximum amount of files allowed. */
+    const MAX_FILES = 50;
+
+    /**
+     * Pulls data from the datasource and adds it to data.
+     *
+     * @param \stdClass $data
+     */
+    public function add_to_data(\stdClass $data) {
+        $files = $this->cms->get_type()->get_images();
+
+        $imagedata = new stdClass();
+        foreach ($files as $file) {
+            $filename = $file->get_filename();
+            $shortfilename = pathinfo($filename, PATHINFO_FILENAME);
+            if ($filename == '.') {
+                continue;
+            }
+            $url = \moodle_url::make_pluginfile_url(
+                context_system::instance()->id,
+                'mod_cms',
+                'cms_type_images',
+                $this->cms->get('typeid'),
+                '/',
+                $filename
+            );
+            $imagedata->$shortfilename = $url->out();
+        }
+        $data->images = $imagedata;
```

Comment:
I think it would be good if there was a method to return the ‘short name’ of the source eg ‘images’ and then the central core managed the appending of all the data into the one place instead of each source writing in with an object pointer

Reply:
This method could also default to returning the name of the class and or component name and so you don’t need to override it

Reply:
Done